RetailCoder.VBE/Navigation/CodeMetrics/CodeMetricsAnalyst.cs

```diff
@@ -0,0 +1,193 @@
+            private List<CodeMetricsResult> moduleResults = new List<CodeMetricsResult>();
+
+            private List<MemberMetricsResult> memberResults = new List<MemberMetricsResult>();
+
+            public CodeMetricsListener(DeclarationFinder finder, IIndenterSettings indenterSettings)
+            {
+                _finder = finder;
+                _indenterSettings = indenterSettings;
+            }
+
+            public override void EnterEndOfLine([NotNull] VBAParser.EndOfLineContext context)
+            {
+                int followingIndentationLevel = 0;
+                // we have a proper newline
+                if (context.NEWLINE() != null)
+                {
+                    // the last whitespace, which is the one in front of the next line's contents
+                    var followingWhitespace = context.whiteSpace().LastOrDefault();
+                    followingIndentationLevel = IndentationLevelFromWhitespace(followingWhitespace);
+                }
+                (currentMember == null ? moduleResults : results).Add(new CodeMetricsResult(1, 0, followingIndentationLevel));
+            }
+
+            public override void EnterIfStmt([NotNull] VBAParser.IfStmtContext context)
+            {
+                results.Add(new CodeMetricsResult(0, 1, 0));
+            }
+
+            public override void EnterElseIfBlock([NotNull] VBAParser.ElseIfBlockContext context)
+            {
+                results.Add(new CodeMetricsResult(0, 1, 0));
+            }
+
+            // notably: NO additional complexity for an Else-Block
+
+            public override void EnterForEachStmt([NotNull] VBAParser.ForEachStmtContext context)
+            {
+                results.Add(new CodeMetricsResult(0, 1, 0));
+            }
+
+            public override void EnterForNextStmt([NotNull] VBAParser.ForNextStmtContext context)
+            {
+                results.Add(new CodeMetricsResult(0, 1, 0));
+            }
+
+            public override void EnterCaseClause([NotNull] VBAParser.CaseClauseContext context)
+            {
+                results.Add(new CodeMetricsResult(0, 1, 0));
+            }
+
+            public override void EnterSubStmt([NotNull] VBAParser.SubStmtContext context)
+            {
+                results.Add(new CodeMetricsResult(0, 1, 0));
+                currentMember = _finder.DeclarationsWithType(DeclarationType.Procedure).Where(d => d.Context == context).First();
+            }
+
+            public override void ExitSubStmt([NotNull] VBAParser.SubStmtContext context)
+            {
+                ExitMeasurableMember();
+            }
+
+            public override void EnterFunctionStmt([NotNull] VBAParser.FunctionStmtContext context)
+            {
+                results.Add(new CodeMetricsResult(0, 1, 0));
+                currentMember = _finder.DeclarationsWithType(DeclarationType.Function).Where(d => d.Context == context).First();
```

Comment:
I think it would be more appropriate to use `UserDeclarations` here. `DeclarationsWithType` is the concatenation of `BuiltInDeclarations` with `UserDeclarations`.